### PR TITLE
Added visitor to validate junction nodes

### PR DESCRIPTION
**Pull request by apmoriarty**
For use when https://github.com/NationalSecurityAgency/datawave/pull/1274 is merged

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/JunctionValidatingVisitor.java [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+    
+    private boolean isValid = true;
+    
+    public static boolean validate(JexlNode script) {
+        JunctionValidatingVisitor visitor = new JunctionValidatingVisitor();
+        script.jjtAccept(visitor, null);
+        return visitor.getIsValid();
+    }
+    
+    public boolean getIsValid() {
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/JunctionValidatingVisitor.java]: Convention here would be `isValid()`

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/JunctionValidatingVisitor.java [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+        return this.isValid;
+    }
+    
+    /*
+     * Pass through
+     */
+    
+    @Override
+    public Object visit(ASTJexlScript node, Object data) {
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/JunctionValidatingVisitor.java]: Same as parent impl.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/JunctionValidatingVisitor.java [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+        if (!isJunctionValid(node)) {
+            return data;
+        }
+        node.childrenAccept(this, data);
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTAndNode node, Object data) {
+        if (!isJunctionValid(node)) {
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/JunctionValidatingVisitor.java]: Just a thought here, let me know what you think.

If I read this method, it's not clear that the isValid flag is getting set unless I jump into `isJunctionValid()`. The `isJunctionValid()` method is doing double-duty. It both checks if the node is valid and has a side-effect. If you broke those two apart, this method could be:

```
public Object visit(ASTAndNode node, Object data) {
  if (!isValid)
    return data;
  else if (!isJunctionValid(node)) {
    isValid = false;
    return data;
  }
  // recurse down tree
}
```

Then `isJunctionValid()` only has a single job of checking a node's validity. It's cool if you wanna skip this but wanted to mention it.

**Review by FineAndDandy — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/JunctionValidatingVisitor.java (hunk 3 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/JunctionValidatingVisitor.java]: Unlike #1299 the validate methods are the same for and/or nodes, going to extract this code to a common method. Good call!

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/JunctionValidatingVisitor.java (hunk 2 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/JunctionValidatingVisitor.java]: Removed in next commit

**Review by jschmidt10 — APPROVED** [review]

**Review by alerman — APPROVED** [review]
LGTM

**Pull request merged**